Detect BC breaks due to parameter name changes, in line with PHP 8.0 named arguments requirements

asgrim opened the pull request:
Fixes #264

asgrim committed:
Added test for parameter name changes becoming a BC break for PHP 8

asgrim committed:
Basic param renames are now a BC break

asgrim commented on src/DetectChanges/BCBreak/FunctionBased/ParameterNameChanged.php:
```diff
@@ -0,0 +1,79 @@
+/**
+ * @todo only apply this to PHP 8+ code
+ *
+ * Detects a change in a parameter type
```
Fix docblocks from C&P

Ocramius commented:
Note: also needs to configure the new rule in `bin/*.php`

asgrim commented on src/DetectChanges/BCBreak/FunctionBased/ParameterNameChanged.php:
```diff
@@ -0,0 +1,79 @@
+     */
+    private function compareParameter(ReflectionParameter $fromParameter, ReflectionParameter $toParameter): iterable
+    {
+        // @todo detect if the method has a @no-named-arguments annotation, and return if so
```
`@todo`

asgrim commented on src/DetectChanges/BCBreak/FunctionBased/ParameterNameChanged.php:
```diff
@@ -0,0 +1,79 @@
+use function Safe\sprintf;
+
+/**
+ * @todo only apply this to PHP 8+ code
```
`@todo`

asgrim replied:
I think the only way we can realistically do this is to:

 - Write a class like `ReadPhpVersionRequirementsFromComposerJson` that:
   - reads `composer.json`
   - find the `require.php` key and stores the min/max value
   - it would be run lazily
 - The class can be constructor-injected into the `ParameterNamedChanged` check
 - The `ParameterNamedChanged` can ask the class if we intersect PHP 8+ at all, if not, return

Ocramius replied:
IMO this also applies to PHP 7.x code - running the same code on 8.x will have these issues, regardless if we're developing on 7.x right now

Ocramius replied:
17adbea4322a7eddbc19f59e14828c5f69eac9a6

asgrim committed:
Improved BC check class doc

asgrim committed:
Added ParameterNameChanged check into bin/rbcc script

asgrim committed:
Support for @no-named-parameters annotation in docblock to suppress parameter name changes being marked as BC breaks

asgrim commented on src/DetectChanges/BCBreak/FunctionBased/ParameterNameChanged.php:
```diff
@@ -0,0 +1,119 @@
+        );
+    }
+
+    private function methodHasNoNamedArgumentsAnnotation(ReflectionFunctionAbstract $function): bool
```
I wonder if any static analysis library will add some kind of PHP attribute like `#[NoNamedAttributes]` or something, we could look for that too, although I think we'd need to add support upstream in Better Reflection for attributes first, so that could be an improvement to come later.

Ocramius replied:
Possibly to be done once the ecosystem agrees on these annotations - too early for us to jump in on this

Ocramius commented on src/DetectChanges/BCBreak/FunctionBased/ParameterNameChanged.php:
```diff
@@ -0,0 +1,119 @@
+     */
+    private function checkSymbols(array $from, array $to): iterable
+    {
+        foreach (array_intersect_key($from, $to) as $index => $commonParameter) {
```
One test is missing here: specifically something like:

```diff
-function foo($a, $b, $c){}
+function foo($b, $c){}
```

This particular BCBreak detector should not pick that up

Ocramius replied:
4fa0de2315627bfaccce94a1aa05dc7b31ef4e34

Ocramius committed:
Removed solved TODO: `ParameterNameChanged` applies regardless of PHP version

Ref: https://github.com/Roave/BackwardCompatibilityCheck/pull/266/files#r472284330

Ocramius committed:
Upgraded `ParameterNameChanged` to comply with latest `roave/better-reflection:5.0.x` changes

This changes types and symbols that are dictated by interface and dependency changes.

Ocramius committed:
Ensuring that added/removed parameters are ignored by the `ParameterNameChanged` detector

Ref: https://github.com/Roave/BackwardCompatibilityCheck/pull/266/files#r497631718

This fixes following escaped mutation:

```
Escaped mutants:
================

1) /app/src/DetectChanges/BCBreak/FunctionBased/ParameterNameChanged.php:87    [M] UnwrapArrayIntersectKey

--- Original
+++ New
@@ @@
      */
     private function checkSymbols(array $from, array $to) : iterable
     {
-        foreach (array_intersect_key($from, $to) as $index => $commonParameter) {
+        foreach ($from as $index => $commonParameter) {
             yield from $this->compareParameter($commonParameter, $to[$index]);
         }
     }
```

Ocramius committed:
Removed redundant namespace declaration in test assets

Ocramius reviewed (approved):
@asgrim I've adapted your patch to latest `6.0.x` state, and it's good to go :ship:

asgrim commented:
Thanks @Ocramius !
